### PR TITLE
storage: use NextPrefix

### PR DESCRIPTION
**Pull request by jbowens**
Update pebbleIterator to use the new NextPrefix method exposed by Pebble Iterators (cockroachdb/pebble#1860). NextPrefix implements semanatics analgous to MVCCIterator.NextKey and is able to make use of low-level optimizations to avoid much of the work of a full seek. Additionally in the case where the very next key is a new MVCC user key, the pebbleMVCCScanner is able to use NextPrefix to avoid a key copy and a key comparison.

```
name                                                                      old speed      new speed       delta
MVCCScan_Pebble/rows=1/versions=1/valueSize=64/numRangeKeys=0-24          6.20MB/s ± 2%   6.24MB/s ± 1%     ~     (p=0.524 n=5+5)
MVCCScan_Pebble/rows=1/versions=1/valueSize=64/numRangeKeys=1-24          3.71MB/s ± 1%   3.90MB/s ± 1%   +5.01%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1/versions=1/valueSize=64/numRangeKeys=100-24         376kB/s ± 2%    670kB/s ± 0%  +78.19%  (p=0.016 n=5+4)
MVCCScan_Pebble/rows=1/versions=2/valueSize=64/numRangeKeys=0-24          4.87MB/s ± 0%   5.00MB/s ± 2%   +2.71%  (p=0.032 n=5+5)
MVCCScan_Pebble/rows=1/versions=2/valueSize=64/numRangeKeys=1-24          3.08MB/s ± 1%   3.18MB/s ± 1%   +3.38%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1/versions=2/valueSize=64/numRangeKeys=100-24         420kB/s ± 0%    758kB/s ± 2%  +80.48%  (p=0.016 n=4+5)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64/numRangeKeys=0-24         3.21MB/s ± 1%   3.27MB/s ± 2%   +1.99%  (p=0.024 n=5+5)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64/numRangeKeys=1-24         2.11MB/s ± 2%   2.23MB/s ± 2%   +5.89%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1/versions=10/valueSize=64/numRangeKeys=100-24        410kB/s ± 2%    688kB/s ± 2%  +67.80%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1/versions=100/valueSize=64/numRangeKeys=0-24        1.55MB/s ± 1%   1.65MB/s ± 2%   +6.19%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1/versions=100/valueSize=64/numRangeKeys=1-24        1.26MB/s ± 2%   1.33MB/s ± 3%   +5.87%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1/versions=100/valueSize=64/numRangeKeys=100-24       460kB/s ± 0%    536kB/s ± 3%  +16.52%  (p=0.016 n=4+5)
MVCCScan_Pebble/rows=10/versions=1/valueSize=64/numRangeKeys=0-24         41.4MB/s ± 2%   41.2MB/s ± 2%     ~     (p=0.516 n=5+5)
MVCCScan_Pebble/rows=10/versions=1/valueSize=64/numRangeKeys=1-24         26.5MB/s ± 2%   27.1MB/s ± 1%   +2.42%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=1/valueSize=64/numRangeKeys=100-24       5.37MB/s ± 2%   6.17MB/s ± 1%  +14.74%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=2/valueSize=64/numRangeKeys=0-24         32.7MB/s ± 2%   31.5MB/s ± 3%   -3.51%  (p=0.032 n=5+5)
MVCCScan_Pebble/rows=10/versions=2/valueSize=64/numRangeKeys=1-24         21.4MB/s ± 0%   21.6MB/s ± 1%   +1.18%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=2/valueSize=64/numRangeKeys=100-24       5.88MB/s ± 1%   6.74MB/s ± 0%  +14.51%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=10/valueSize=64/numRangeKeys=0-24        17.2MB/s ± 2%   18.5MB/s ± 1%   +7.73%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=10/valueSize=64/numRangeKeys=1-24        11.6MB/s ± 2%   13.2MB/s ± 2%  +13.93%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=10/valueSize=64/numRangeKeys=100-24      4.56MB/s ± 1%   5.58MB/s ± 1%  +22.36%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=100/valueSize=64/numRangeKeys=0-24       6.79MB/s ± 3%   7.50MB/s ± 2%  +10.45%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=100/valueSize=64/numRangeKeys=1-24       5.53MB/s ± 2%   6.31MB/s ± 2%  +13.95%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10/versions=100/valueSize=64/numRangeKeys=100-24     2.94MB/s ± 1%   3.54MB/s ± 1%  +20.27%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64/numRangeKeys=0-24         125MB/s ± 1%    124MB/s ± 1%     ~     (p=0.548 n=5+5)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64/numRangeKeys=1-24        84.2MB/s ± 3%   86.1MB/s ± 2%     ~     (p=0.222 n=5+5)
MVCCScan_Pebble/rows=100/versions=1/valueSize=64/numRangeKeys=100-24      33.7MB/s ± 3%   38.0MB/s ± 1%  +12.83%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=2/valueSize=64/numRangeKeys=0-24        92.7MB/s ± 1%   94.5MB/s ± 1%   +1.90%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=2/valueSize=64/numRangeKeys=1-24        62.5MB/s ± 1%   65.4MB/s ± 1%   +4.75%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=2/valueSize=64/numRangeKeys=100-24      32.0MB/s ± 1%   36.4MB/s ± 1%  +13.51%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64/numRangeKeys=0-24       36.5MB/s ± 3%   41.3MB/s ± 4%  +13.05%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64/numRangeKeys=1-24       23.7MB/s ± 2%   30.8MB/s ± 2%  +29.93%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=10/valueSize=64/numRangeKeys=100-24     15.8MB/s ± 1%   21.3MB/s ± 2%  +34.67%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=100/valueSize=64/numRangeKeys=0-24      11.1MB/s ± 1%   12.6MB/s ± 2%  +13.08%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=100/valueSize=64/numRangeKeys=1-24      9.06MB/s ± 2%  11.08MB/s ± 2%  +22.29%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=100/versions=100/valueSize=64/numRangeKeys=100-24    6.65MB/s ± 1%   8.91MB/s ± 2%  +33.95%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=1/valueSize=64/numRangeKeys=0-24        175MB/s ± 2%    176MB/s ± 2%     ~     (p=0.421 n=5+5)
MVCCScan_Pebble/rows=1000/versions=1/valueSize=64/numRangeKeys=1-24        120MB/s ± 1%    124MB/s ± 1%   +3.59%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=1/valueSize=64/numRangeKeys=100-24     81.3MB/s ± 2%   87.5MB/s ± 1%   +7.62%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=2/valueSize=64/numRangeKeys=0-24        122MB/s ± 1%    132MB/s ± 2%   +7.88%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=2/valueSize=64/numRangeKeys=1-24       82.5MB/s ± 2%   90.5MB/s ± 2%   +9.70%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=2/valueSize=64/numRangeKeys=100-24     65.2MB/s ± 1%   72.0MB/s ± 1%  +10.41%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=10/valueSize=64/numRangeKeys=0-24      42.9MB/s ± 1%   50.2MB/s ± 2%  +17.02%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=10/valueSize=64/numRangeKeys=1-24      27.3MB/s ± 1%   36.7MB/s ± 1%  +34.49%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=10/valueSize=64/numRangeKeys=100-24    21.4MB/s ± 3%   31.4MB/s ± 4%  +46.71%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=100/valueSize=64/numRangeKeys=0-24     12.5MB/s ± 3%   14.6MB/s ± 5%  +16.55%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=100/valueSize=64/numRangeKeys=1-24     10.3MB/s ± 4%   13.2MB/s ± 2%  +28.04%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=1000/versions=100/valueSize=64/numRangeKeys=100-24   8.00MB/s ± 4%  11.47MB/s ± 4%  +43.25%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=1/valueSize=64/numRangeKeys=0-24       193MB/s ± 2%    197MB/s ± 1%   +2.33%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=1/valueSize=64/numRangeKeys=1-24       130MB/s ± 2%    133MB/s ± 1%   +2.05%  (p=0.032 n=5+5)
MVCCScan_Pebble/rows=10000/versions=1/valueSize=64/numRangeKeys=100-24     104MB/s ± 2%    108MB/s ± 1%   +3.41%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=2/valueSize=64/numRangeKeys=0-24       137MB/s ± 1%    143MB/s ± 2%   +4.14%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=2/valueSize=64/numRangeKeys=1-24      89.4MB/s ± 2%   95.8MB/s ± 0%   +7.10%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=2/valueSize=64/numRangeKeys=100-24    77.3MB/s ± 2%   82.3MB/s ± 2%   +6.47%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=10/valueSize=64/numRangeKeys=0-24     44.9MB/s ± 1%   52.7MB/s ± 3%  +17.30%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=10/valueSize=64/numRangeKeys=1-24     28.6MB/s ± 2%   39.0MB/s ± 5%  +36.19%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=10/valueSize=64/numRangeKeys=100-24   22.6MB/s ± 7%   34.5MB/s ± 5%  +52.68%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=100/valueSize=64/numRangeKeys=0-24    13.2MB/s ± 3%   15.2MB/s ± 4%  +15.51%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=100/valueSize=64/numRangeKeys=1-24    10.5MB/s ± 4%   13.4MB/s ± 2%  +27.11%  (p=0.008 n=5+5)
MVCCScan_Pebble/rows=10000/versions=100/valueSize=64/numRangeKeys=100-24  8.13MB/s ± 6%  11.87MB/s ± 3%  +45.98%  (p=0.008 n=5+5)
```

Epic: None
Informs #83049.
Release note: None

### CODE REVIEW

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1443 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJLjzoj6mTiViCYGoBc:-NJLjzoj6mTiViCYGoBd:b-2lwizb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_mvcc_scanner.go#L1443)):*
> ```Go
> 			p.parent.Next()
> 			return p.updateCurrent()
> 		}
> ```

I removed this, because as far as I know, it's unnecessary. If a `*pebble.Iterator` is peeked off the beginning of iteration, `Next`-ing it is fine and will already handle seeking to the start bound.

**Review by erikgrinaker — APPROVED** [review]
Very nice! Do you have an `MVCCScan` benchmark comparison?

> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-:-NJPcNt01Pr_dCKx5sed:bl3kqm5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @sumeerbhola)

___
*[`pkg/storage/pebble_iterator.go` line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJPcj6ZCszgJg81dQhB:-NJPcj6ZCszgJg81dQhC:b1yvif5) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_iterator.go#L471)):*
> ```Go
> 		return
> 	}
> 	p.keyBuf = append(p.keyBuf[:0], p.UnsafeKey().Key...)
> ```

This kind of messes up the `if p.prefix` branch below, because it assumes `p.keyBuf` already contains the user key. We should move the copy into that branch. I don't think it has any practical consequence though, because we'll do a prefix seek to a `MinTimestamp` version of an arbitrary key and exhaust the iterator, but better safe than sorry.

Should we move the `p.prefix` logic into Pebble entirely, such that calling `NextPrefix()` after a `SeekPrefixGE` call exhausts the iterator?
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 960 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJPeK8z6xTb2riiEMUH:-NJPeK8z6xTb2riiEMUI:b-gunp5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b715f14bcc2d8b116f905119090e42ff8d967f42/pkg/storage/pebble_mvcc_scanner.go#L960)):*
> ```Go
> func (p *pebbleMVCCScanner) nextKey() bool {
> 	// MVCCIterator.NextKey may not be called while in the reverse direction.
> 	if p.parentReverse {
> ```

I'm getting rid of `parentReverse` over in #93753. Do we need this check? Alternatively, can we move this down into the Pebble iterator and expose it as `IsReverse()`?
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1443 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJLjzoj6mTiViCYGoBc:-NJPf-8L2WoHzI6yRWCI:bb18qvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_mvcc_scanner.go#L1443)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I removed this, because as far as I know, it's unnecessary. If a `*pebble.Iterator` is peeked off the beginning of iteration, `Next`-ing it is fine and will already handle seeking to the start bound.
</blockquote></details>

That sounds right.

**Review by jbowens — COMMENTED** [review]
The MVCCScan benchmark numbers from cockroachdb/pebble#1860 were collected with this change. I'll duplicate them here for posterity.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-:-NJQooiV-ioNjlcy5FKl:bloggn8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/storage/pebble_iterator.go` line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJPcj6ZCszgJg81dQhB:-NJQqhz_8OhAucQhItjV:bud7lep) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_iterator.go#L471)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This kind of messes up the `if p.prefix` branch below, because it assumes `p.keyBuf` already contains the user key. We should move the copy into that branch. I don't think it has any practical consequence though, because we'll do a prefix seek to a `MinTimestamp` version of an arbitrary key and exhaust the iterator, but better safe than sorry.

Should we move the `p.prefix` logic into Pebble entirely, such that calling `NextPrefix()` after a `SeekPrefixGE` call exhausts the iterator?
</blockquote></details>

Good call. Created cockroachdb/pebble#2194.
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 960 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJPeK8z6xTb2riiEMUH:-NJQqtWz3HPyPX_44Cnz:b2sm9rt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b715f14bcc2d8b116f905119090e42ff8d967f42/pkg/storage/pebble_mvcc_scanner.go#L960)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I'm getting rid of `parentReverse` over in #93753. Do we need this check? Alternatively, can we move this down into the Pebble iterator and expose it as `IsReverse()`?
</blockquote></details>

I'll take a look at #93753—but yeah, I think we need some form of this. The result of calling `NextPrefix` on an iterator oriented in the reverse direction is undefined. I can take a second look to see how feasible it is to extend `NextPrefix` to direction changes.

**Review by erikgrinaker — APPROVED** [review]
> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @sumeerbhola)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 960 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJPeK8z6xTb2riiEMUH:-NJdoAhJ-Ci0Vurgn8Se:b-n0pcp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b715f14bcc2d8b116f905119090e42ff8d967f42/pkg/storage/pebble_mvcc_scanner.go#L960)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I'll take a look at #93753—but yeah, I think we need some form of this. The result of calling `NextPrefix` on an iterator oriented in the reverse direction is undefined. I can take a second look to see how feasible it is to extend `NextPrefix` to direction changes.
</blockquote></details>

Ok, let me know if I should roll back the `parentReverse` removal there.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
:lgtm: 

> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-:-NJgVo655awkmXGy2A0n:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/storage/pebble_iterator.go` line 471 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJgURjm-h0VCwFFblMl:-NJgURjm-h0VCwFFblMm:b-9rp77c) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c0ef717c2c7ccce1b48dc0f8451573ca093d77/pkg/storage/pebble_iterator.go#L471)):*
> ```Go
> 	// propagate RangeKeyChanged() correctly.
> 	if p.prefix {
> 		// Seek to the latest possible key for this prefix, exhausting iter.
> ```

Will this if-block ever be taken in production code?
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 960 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJgVhrwCy0o5ZcArqIW:-NJgVhrwCy0o5ZcArqIX:b-apjia7) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c0ef717c2c7ccce1b48dc0f8451573ca093d77/pkg/storage/pebble_mvcc_scanner.go#L960)):*
> ```Go
> func (p *pebbleMVCCScanner) nextKey() bool {
> 	// MVCCIterator.NextKey may not be called while in the reverse direction.
> 	if p.parentReverse {
> ```

I didn't look at this `parentReverse` block yet. I'll look at @erikgrinaker's PR first.
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJgV5E8179Uy_p2TSh1:-NJgV5E8179Uy_p2TSh2:bhx4wtp) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_mvcc_scanner.go#L1437)):*
> ```Go
> 			// We were peeked off the beginning of iteration. Seek to the first
> 			// entry, and then advance one step.
> 			p.parent.SeekGE(MVCCKey{Key: p.start})
> ```

Is this an unrelated change?
Which uses the fact that an iter that has prev'd to exhaustion can be nexted to get the first entry?
May be worth a code comment pointing out this case and that it does not need special casing here.

**Comment by erikgrinaker** [comment]
Would appreciate if we could hold this until I get a fix in for #94015. I need something I can backport to 22.2, and that will also add more comprehensive `RangeKeyChanged()` tests that need to pass for this PR before merging.

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/storage/pebble_iterator.go` line 471 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJgURjm-h0VCwFFblMl:-NKJ7iH3EZ-gSYZHLnua:bmnj0ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c0ef717c2c7ccce1b48dc0f8451573ca093d77/pkg/storage/pebble_iterator.go#L471)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Will this if-block ever be taken in production code?
</blockquote></details>

I don't believe so. Also, removed this if-block now that NextPrefix during prefix-iteration mode exhausts the iterator (cockroachdb/pebble#2194).
___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJgV5E8179Uy_p2TSh1:-NKJ8npu2iNu30ct76v8:boqpt08) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_mvcc_scanner.go#L1437)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this an unrelated change?
Which uses the fact that an iter that has prev'd to exhaustion can be nexted to get the first entry?
May be worth a code comment pointing out this case and that it does not need special casing here.
</blockquote></details>

yeah, it was just something I noticed. Added a code comment

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 960 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJPeK8z6xTb2riiEMUH:-NKJWidxA-dOP76un693:b-vk4sog) ([raw file](https://github.com/cockroachdb/cockroach/blob/b715f14bcc2d8b116f905119090e42ff8d967f42/pkg/storage/pebble_mvcc_scanner.go#L960)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Ok, let me know if I should roll back the `parentReverse` removal there.
</blockquote></details>

Put up cockroachdb/pebble#3022—we can nix `parentReverse`

**Review by sumeerbhola — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @jbowens)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 960 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJPeK8z6xTb2riiEMUH:-NKxXceL1oteMyvfHJOj:b-gymsyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b715f14bcc2d8b116f905119090e42ff8d967f42/pkg/storage/pebble_mvcc_scanner.go#L960)):*
> cockroachdb/pebble#3022

Couldn't find this.
Also, the `parentReverse` code is still here. Should I hold off looking at this code?

**Review by erikgrinaker — APPROVED** [review]
Lovely simplification, thanks!

**Review by jbowens — COMMENTED** [review]
Unfortunately, there's been quite a bit of churn in this code recently, and it looks like something broke. I'll look into it but might not get to it until next week.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-:-NL2nPdu17QTsTW7Qq4X:b-grfo8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @sumeerbhola)

**Review by jbowens — COMMENTED** [review]
> Unfortunately, there's been quite a bit of churn in this code recently, and it looks like something broke. I'll look into it but might not get to it until next week.

I was missing a call to `processRangeKeys`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-:-NL30ul78t3liF0AjdQM:b-101npg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @sumeerbhola)

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @jbowens)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJgV5E8179Uy_p2TSh1:-NL3gXqc9tt7ERKggsDq:bbmyquh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_mvcc_scanner.go#L1437)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

yeah, it was just something I noticed. Added a code comment
</blockquote></details>

Dropped this change to avoid the `SeekGE`?

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @sumeerbhola)

___
*[`pkg/storage/pebble_mvcc_scanner.go` line 1437 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-NJgV5E8179Uy_p2TSh1:-NL3tgT84bcNpALjBnrQ:b-a3ukmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbea1957a83cd89f412e3c487ca6ac1fde82cc65/pkg/storage/pebble_mvcc_scanner.go#L1437)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Dropped this change to avoid the `SeekGE`?
</blockquote></details>

I dropped it when the rebase wasn't smooth, but it was simple to add it back in.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93723#-:-NL6YuH6Dlrzg2ricMBI:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @jbowens)

**Comment by jbowens** [comment]
tftrs!

bors r+

**Pull request merged**
